core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java

```diff
@@ -248,14 +294,22 @@ public SELF withExposedService(String serviceName, int servicePort) {
 
         // Ambassador container will be started together after docker compose has started
         int ambassadorPort = nextAmbassadorPort.getAndIncrement();
-        ambassadorPortMappings.computeIfAbsent(serviceName, __ -> new ConcurrentHashMap<>()).put(servicePort, ambassadorPort);
-        ambassadorContainer.withTarget(ambassadorPort, serviceName, servicePort);
-        ambassadorContainer.addLink(new FutureContainer(this.project + "_" + serviceName), serviceName);
-        return self();
-    }
+        ambassadorPortMappings.computeIfAbsent(serviceInstanceName, __ -> new ConcurrentHashMap<>()).put(servicePort, ambassadorPort);
+        ambassadorContainer.withTarget(ambassadorPort, serviceInstanceName, servicePort);
+        ambassadorContainer.addLink(new FutureContainer(this.project + "_" + serviceInstanceName), serviceInstanceName);
 
-    public DockerComposeContainer withExposedService(String serviceName, int instance, int servicePort) {
-        return withExposedService(serviceName + "_" + instance, servicePort);
+        /*
+         * can have multiple wait strategies for a single container, e.g. if waiting on several ports
+         * if no wait strategy is defined, the WaitAllStrategy will return immediately
+         */
+        final WaitAllStrategy waitAllStrategy = waitStrategyMap.computeIfAbsent(serviceInstanceName, __ ->
+            (WaitAllStrategy) new WaitAllStrategy().withStartupTimeout(startupTimeout));
+
+        if(waitStrategy != null) {
+            waitStrategy.withStartupTimeout(startupTimeout);
```

Comment:
already set in `computeIfAbsent`

Reply:
The `startupTimeout` is set on the `WaitAllStrategy` that wraps the (potentially) multiple `WaitStrategy` instances for a particular service, but I thought it would be a good idea to apply the same timeout to all of those strategies.

Otherwise you could have the case where one strategy could time out after 1 minute, but the `WaitAllStrategy` could be set to 5 minutes.

As an example:
```java
ew DockerComposeContainer(new File("src/test/resources/compose-test.yml"))
            .withStartupTimeout(Duration.ofSeconds(120))
            .withExposedService("redis_1", REDIS_PORT, Wait.forListeningPort())
            .withExposedService("redis_1", SOME_OTHER_PORT, Wait.forListeningPort());
```
Two instances of `HostPortWaitStrategy` would be applied to the same service here. The idea was to apply the same timeout to both.

WDYT? Does it make sense to do this?

Reply:
I see. Hm, what if we rename the method to `withDefaultStartupTimeout` here?
Not sure tho.

We need to be careful because of existing "startup timeout" in GenericContainer where it applies to the container and not a set of containers.

Reply:
I had moved `SELF withStartupTimeout(Duration startupTimeout);` to a new interface `StartupTImeout`, which I thought was inherited by `Container` and implemented by 'GenericContainer` and `DockerComposeContainer`, but part of that seems to have been lost in one of the refactors.

Currently that interface is only implemented by `DockerComposeContainer`, while `Container` has a method with the same signature. 
I could remove that interface and just add the method directly to `DockerComposeContainer`? Or add a method like `withDefaultStartupTimeout`, as you mentioned?

In general though, should we be applying a single timeout to all `WaitStrategy` instances on docker-compose?
`AbstractWaitStrategy` has a default timeout of 60s, so it would be difficult to check if a default should be applied to all strategies, or if it has been overridden for a particular strategy.

Since there's probably multiple services involved, the timeouts are likely to need to be longer than for a single `GenericContainer`.

Reply:
I would remove the interface, at least I see no use cases for it

Re "single timeout to all WaitStrategy instances" - good point about "all strategies". Maybe we should consider removing the method (`withStartupTimeout`) at all?

So that users will explicitly add the timeout per exposed service:
```java
.withExposedService("redis_1", 6379, Wait.forHttp("/").withStartupTimeout(...))
```

Reply:
There's still the case where you may have more than one `WaitStrategy` per service, e.g., it if exposes two ports

```java
.withExposedService("redis_1", 6379, Wait.forListeningPort())
.withExposedService("redis_1", 6380, Wait.forListeningPort())
```

I think each service needs to have a `WaitAllStrategy` to wrap the possible multiple strategies, and that also needs to have a timout. Or at least, it has one in the current implementation

Reply:
`WaitAllStrategy` can have an infinite duration (means the duration will be controlled by inner waiters):
```java
.withExposedService("redis_1", 6379, Wait.forListeningPort().withStartupTimeout(timeout1))
.withExposedService("redis_1", 6380, Wait.forListeningPort()) // Uses default
```

no?

Reply:
The current implementation of `WaitAllStrategy` has a default timeout of 30 seconds. For this case we can either implement a new `InfiniteTimeWaitAllStrategy`, or just set the timeout for an improbably long time for the `WaitAllStrategy` created in `DockerComposeContainer` per service. Say 30 minutes or so?

Reply:
sounds good